// @umijs/preset-umi/dist/commands/dev/createRouteMiddleware.js
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// createRouteMiddleware.ts
var createRouteMiddleware_exports = {};
__export(createRouteMiddleware_exports, {
  createRouteMiddleware: () => createRouteMiddleware
});
module.exports = __toCommonJS(createRouteMiddleware_exports);
var import_server = require("@umijs/server");
var import_getAssetsMap = require("./getAssetsMap");
var import_getMarkupArgs = require("./getMarkupArgs");
function createRouteMiddleware(opts) {
  return ({ compiler }) => {
    let webpackStats = null;
    let onStats = null;
    compiler.hooks.done.tap("umiRouteMiddleware", (stats) => {
      webpackStats = stats;
      onStats == null ? void 0 : onStats(stats);
    });
    async function getStats() {
      if (webpackStats)
        return Promise.resolve(webpackStats);
      return new Promise((resolve) => {
        onStats = (stats) => {
          resolve(stats);
        };
      });
    }
    return async (req, res, next) => {
      var _a;
      const markupArgs = await (0, import_getMarkupArgs.getMarkupArgs)(opts);
      let assetsMap = {};
      const stats = await getStats();
      assetsMap = (0, import_getAssetsMap.getAssetsMap)({
        stats,
        publicPath: opts.api.config.publicPath
      });
      const requestHandler = await (0, import_server.createRequestHandler)(__spreadProps(__spreadValues({}, markupArgs), {
        styles: markupArgs.styles.concat(assetsMap["umi.css"] || []),
        scripts: (assetsMap["umi.js"] || []).concat(markupArgs.scripts),
        esmScript: false,
        historyType: ((_a = opts.api.config.history) == null ? void 0 : _a.type) || "browser"
      }));
      requestHandler(req, res, next);
    };
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  createRouteMiddleware
});
