// @umijs/preset-umi/dist/commands/dev/getBabelOpts.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target, mod));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// getBabelOpts.ts
var getBabelOpts_exports = {};
__export(getBabelOpts_exports, {
  getBabelOpts: () => getBabelOpts
});
module.exports = __toCommonJS(getBabelOpts_exports);
var import_utils = require("@umijs/utils");
async function getBabelOpts(opts) {
  const isGTEReact17 = import_utils.semver.gte(opts.api.appData.react.version, "17.0.0");
  const babelPresetOpts = await opts.api.applyPlugins({
    key: "modifyBabelPresetOpts",
    initialValue: {
      presetEnv: {},
      presetReact: __spreadValues({
        runtime: isGTEReact17 ? "automatic" : "classic"
      }, isGTEReact17 ? {} : { importSource: void 0 }),
      presetTypeScript: {},
      pluginTransformRuntime: {},
      pluginLockCoreJS: {},
      pluginDynamicImportNode: false,
      pluginAutoCSSModules: opts.api.config.autoCSSModules
    }
  });
  const babelPreset = [
    require.resolve("@umijs/babel-preset-umi"),
    babelPresetOpts
  ];
  const extraBabelPresets = await opts.api.applyPlugins({
    key: "addExtraBabelPresets",
    initialValue: []
  });
  const extraBabelPlugins = await opts.api.applyPlugins({
    key: "addExtraBabelPlugins",
    initialValue: []
  });
  const beforeBabelPresets = await opts.api.applyPlugins({
    key: "addBeforeBabelPresets",
    initialValue: []
  });
  const beforeBabelPlugins = await opts.api.applyPlugins({
    key: "addBeforeBabelPlugins",
    initialValue: []
  });
  return {
    babelPreset,
    extraBabelPlugins,
    extraBabelPresets,
    beforeBabelPresets,
    beforeBabelPlugins
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  getBabelOpts
});
