// @umijs/preset-umi/dist/commands/dev/watch.js
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// watch.ts
var watch_exports = {};
__export(watch_exports, {
  addUnWatch: () => addUnWatch,
  createDebouncedHandler: () => createDebouncedHandler,
  expandCSSPaths: () => expandCSSPaths,
  expandJSPaths: () => expandJSPaths,
  unwatch: () => unwatch,
  watch: () => watch
});
module.exports = __toCommonJS(watch_exports);
var import_utils = require("@umijs/utils");
var unWatches = [];
function addUnWatch(unWatcher) {
  unWatches.push(unWatcher);
}
function watch(opts) {
  const watcher = import_utils.chokidar.watch(opts.path, __spreadProps(__spreadValues({}, opts.watchOpts), {
    ignoreInitial: true
  }));
  watcher.on("all", opts.onChange);
  if (opts.addToUnWatches) {
    addUnWatch(() => {
      watcher.close();
    });
  }
  return watcher;
}
function createDebouncedHandler(opts) {
  let timer = null;
  let files = [];
  return (event, path) => {
    if (timer) {
      clearTimeout(timer);
    }
    files.push({ event, path: (0, import_utils.winPath)(path) });
    timer = setTimeout(async () => {
      timer = null;
      await opts.onChange({ files });
      files = [];
    }, opts.timeout || 2e3);
  };
}
function unwatch() {
  unWatches.forEach((unWatch) => unWatch());
}
function expandJSPaths(path) {
  return [".js", ".jsx", ".ts", ".tsx"].map((ext) => {
    return `${path}${ext}`;
  });
}
function expandCSSPaths(path) {
  return [".css", ".less", ".scss", ".sass"].map((ext) => {
    return `${path}${ext}`;
  });
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  addUnWatch,
  createDebouncedHandler,
  expandCSSPaths,
  expandJSPaths,
  unwatch,
  watch
});
