// @umijs/preset-umi/dist/features/vite/vite.js
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// vite.ts
var vite_exports = {};
__export(vite_exports, {
  default: () => vite_default
});
module.exports = __toCommonJS(vite_exports);
var vite_default = (api) => {
  api.describe({
    key: "vite",
    config: {
      schema(Joi) {
        return Joi.object();
      }
    },
    enableBy: api.EnableBy.config
  });
  api.modifyConfig((memo) => {
    memo.alias["@fs"] = api.cwd;
    return memo;
  });
  api.register({
    key: "onBeforeCompiler",
    stage: Number.POSITIVE_INFINITY,
    async fn() {
      await api.applyPlugins({
        key: "updateAppDataDeps",
        type: api.ApplyPluginsType.event
      });
    }
  });
  api.modifyViteConfig((memo) => {
    var _a, _b;
    memo.optimizeDeps = __spreadProps(__spreadValues({}, memo.optimizeDeps || {}), {
      include: (_b = (_a = memo.optimizeDeps) == null ? void 0 : _a.include) == null ? void 0 : _b.concat(Object.values(api.appData.deps).map(({ matches }) => matches[0]).filter((item) => (item == null ? void 0 : item.startsWith("@fs")) && !(item == null ? void 0 : item.includes("node_modules"))))
    });
    return memo;
  });
  let buildStats;
  api.onBuildComplete(({ err, stats }) => {
    if (!err) {
      buildStats = stats;
    }
  });
  api.modifyHTML(($) => {
    if (buildStats) {
      $("head").append(buildStats.extraHtml.head);
      $("body").append(buildStats.extraHtml.body);
    }
    return $;
  });
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {});
