// @umijs/renderer-react/dist/withRouter.js
import _objectSpread from "@babel/runtime/helpers/objectSpread2";
import React from 'react';
import { useLocation, useNavigate, useParams } from 'react-router-dom';
import { jsx as _jsx } from "react/jsx-runtime";
export function withRouter(Component) {
  function ComponentWithRouterProp(props) {
    var location = useLocation();
    var navigate = useNavigate();
    var params = useParams();
    var match = {
      params: params
    };
    var history = {
      back: function back() {
        return navigate(-1);
      },
      goBack: function goBack() {
        return navigate(-1);
      },
      location: location,
      push: function push(url, state) {
        return navigate(url, {
          state: state
        });
      },
      replace: function replace(url, state) {
        return navigate(url, {
          replace: true,
          state: state
        });
      }
    };
    return /*#__PURE__*/_jsx(Component, _objectSpread(_objectSpread({}, props), {}, {
      history: history,
      location: location,
      match: match,
      params: params,
      navigate: navigate
    }));
  }

  return ComponentWithRouterProp;
}
