// @umijs/mfsu/dist/dep/dep.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target, mod));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// dep.ts
var dep_exports = {};
__export(dep_exports, {
  Dep: () => Dep
});
module.exports = __toCommonJS(dep_exports);
var import_utils = require("@umijs/utils");
var import_assert = __toESM(require("assert"));
var import_enhanced_resolve = __toESM(require("enhanced-resolve"));
var import_fs = require("fs");
var import_path = require("path");
var import_constants = require("../constants");
var import_trimFileContent = require("../utils/trimFileContent");
var import_getExposeFromContent = require("./getExposeFromContent");
var resolver = import_enhanced_resolve.default.create({
  mainFields: ["module", "browser", "main"],
  extensions: [".wasm", ".mjs", ".js", ".jsx", ".ts", ".tsx", ".json"],
  exportsFields: ["exports"],
  conditionNames: ["import", "module", "require", "node"]
});
async function resolve(context, path) {
  return new Promise((resolve2, reject) => {
    resolver(context, path, (err, result) => err ? reject(err) : resolve2(result));
  });
}
var Dep = class {
  constructor(opts) {
    this.file = (0, import_utils.winPath)(opts.file);
    this.version = opts.version;
    this.cwd = opts.cwd;
    this.shortFile = this.file;
    this.normalizedFile = this.shortFile.replace(/\//g, "_").replace(/:/g, "_");
    this.filePath = `${import_constants.MF_VA_PREFIX}${this.normalizedFile}.js`;
    this.mfsu = opts.mfsu;
  }
  async buildExposeContent() {
    const isNodeNatives = !!process.binding("natives")[this.file];
    if (isNodeNatives) {
      return (0, import_trimFileContent.trimFileContent)(this.mfsu.opts.excludeNodeNatives ? `
const _ = require('${this.file}');
module.exports = _;
      ` : `
import _ from '${this.file}';
export default _;
export * from '${this.file}';
      `);
    }
    const realFile = await this.getRealFile();
    (0, import_assert.default)(realFile, `filePath not found of ${this.file}`);
    const content = (0, import_fs.readFileSync)(realFile, "utf-8");
    return await (0, import_getExposeFromContent.getExposeFromContent)({
      content,
      filePath: realFile,
      dep: this
    });
  }
  async getRealFile() {
    try {
      return await resolve(this.cwd, this.file);
    } catch (e) {
      return null;
    }
  }
  static buildDeps(opts) {
    return Object.keys(opts.deps).map((file) => {
      return new Dep(__spreadProps(__spreadValues({}, opts.deps[file]), {
        cwd: opts.cwd,
        mfsu: opts.mfsu
      }));
    });
  }
  static getDepVersion(opts) {
    if (!!process.binding("natives")[opts.dep]) {
      return "*";
    }
    const dep = (0, import_path.isAbsolute)(opts.dep) ? opts.dep : (0, import_path.join)(opts.cwd, "node_modules", opts.dep);
    const pkg = import_utils.pkgUp.pkgUpSync({
      cwd: dep
    });
    (0, import_assert.default)(pkg, `package.json not found for ${opts.dep}`);
    return require(pkg).version || null;
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Dep
});
